cmd/frontend/graphqlbackend/search_results.go

```diff
@@ -520,95 +520,98 @@ func (r *searchResolver) evaluateLeaf(ctx context.Context) (_ *SearchResultsReso
-			}
-			continue
-		}
-		merged = append(merged, r)
+func (d *searchResultDeduper) init() {
+	if d.seenFileMatches == nil {
+		d.seenFileMatches = make(map[string]*FileMatchResolver)
+		d.seenRepoMatches = make(map[string]*RepositoryResolver)
+		d.seenCommitMatches = make(map[string]*CommitSearchResultResolver)
+		d.seenDiffMatches = make(map[string]*CommitSearchResultResolver)
 	}
+}
 
-	for _, leftMatch := range left.SearchResults {
-		if leftFileMatch, ok := leftMatch.ToFileMatch(); ok {
-			rightFileMatch := rightFileMatches[leftFileMatch.uri]
-			if rightFileMatch == nil {
-				// no overlap with existing matches.
-				merged = append(merged, leftMatch)
-				count++
-				continue
-			}
-			// merge line matches with a file match that already exists.
-			rightFileMatch.appendMatches(leftFileMatch)
-			rightFileMatches[leftFileMatch.uri] = rightFileMatch
-			continue
+// Add adds a SearchResultResolver to the deduper, merging it into
+// a previously added SearchResultResolver if the URL has already been seen
+func (d *searchResultDeduper) Add(r SearchResultResolver) {
+	d.init()
+
+	if fileMatch, ok := r.ToFileMatch(); ok {
+		if prev, seen := d.seenFileMatches[fileMatch.uri]; seen {
+			prev.appendMatches(fileMatch)
+			return
 		}
+		d.seenFileMatches[fileMatch.uri] = fileMatch
+		return
+	}
 
-		if leftRepoMatch, ok := leftMatch.ToRepository(); ok {
-			rightRepoMatch := rightRepoMatches[string(leftRepoMatch.URL())]
-			if rightRepoMatch == nil {
-				// no overlap with existing matches.
-				merged = append(merged, leftMatch)
-				count++
-			}
-			continue
+	if repoMatch, ok := r.ToRepository(); ok {
+		if _, seen := d.seenRepoMatches[repoMatch.URL()]; seen {
+			return
 		}
+		d.seenRepoMatches[repoMatch.URL()] = repoMatch
+		return
+	}
 
-		if leftCommitMatch, ok := leftMatch.ToCommitSearchResult(); ok {
-			if leftCommitMatch.DiffPreview() != nil {
-				rightDiffMatch := rightDiffMatches[leftCommitMatch.URL()]
-				if rightDiffMatch == nil {
-					merged = append(merged, leftCommitMatch)
-					count++
-				}
-			} else {
-				rightCommitMatch := rightCommitMatches[leftCommitMatch.URL()]
-				if rightCommitMatch == nil {
-					merged = append(merged, leftCommitMatch)
-					count++
-				}
+	if commitMatch, ok := r.ToCommitSearchResult(); ok {
+		if commitMatch.DiffPreview() != nil {
+			if _, seen := d.seenDiffMatches[commitMatch.URL()]; seen {
+				return
 			}
-			continue
+			d.seenDiffMatches[commitMatch.URL()] = commitMatch
+			return
+		} else {
+			if _, seen := d.seenCommitMatches[commitMatch.URL()]; seen {
+				return
+			}
+			d.seenCommitMatches[commitMatch.URL()] = commitMatch
+			return
 		}
-		merged = append(merged, leftMatch)
 	}
 
-	for _, v := range rightFileMatches {
-		merged = append(merged, v)
+	return
+}
+
+// Deduped returns a slice of SearchResultResolvers, deduplicated from
+// the SearchResultResolvers that were added with Add
+func (d *searchResultDeduper) Deduped() []SearchResultResolver {
```

Comment:
```suggestion
// Dedupe deduplicates SearchResultResolvers that were added with Add.
func (d *searchResultDeduper) Dedupe() []SearchResultResolver {
```

Nit: would just make this a verb.